pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CaseTransformFunction.java

```diff
@@ -241,226 +266,768 @@ public TransformResultMetadata getResultMetadata() {
+      for (int docId = unselectedDocs.nextSetBit(0); docId >= 0; docId = unselectedDocs.nextSetBit(docId + 1)) {
+        if (conditions[docId] == 1) {
+          unselectedDocs.clear(docId);
+          _selectedResults[docId] = i;
+        }
+      }
+      if (unselectedDocs.isEmpty()) {
+        break;
       }
     }
     // try to prune clauses now
     for (int i = 0; i < numDocs; i++) {
-      _selections[_selectedResults[i]] = true;
-    }
-    int numSelections = 0;
-    for (boolean selection : _selections) {
-      if (selection) {
-        numSelections++;
+      if (_selectedResults[i] != -1) {
+        _selections[_selectedResults[i]] = true;
       }
     }
-    _numSelections = numSelections;
     return _selectedResults;
   }
 
+  // Returns an array of valueBlock length to indicate whether a row is selected or not.
+  // When nullHandlingEnabled is set to true, we also check whether the row is null and set to false if null.
+  private static int[] getWhenConditions(TransformFunction whenStatement, ValueBlock valueBlock,
+      boolean nullHandlingEnabled) {
+    if (!nullHandlingEnabled) {
+      return whenStatement.transformToIntValuesSV(valueBlock);
+    }
+    Pair<int[], RoaringBitmap> result = whenStatement.transformToIntValuesSVWithNull(valueBlock);
+    RoaringBitmap bitmap = result.getRight();
+    int[] intResult = result.getLeft();
+    if (bitmap != null) {
+      bitmap.forEach((IntConsumer) (i) -> {
+        if (bitmap.contains(i)) {
+          intResult[i] = 0;
+        }
+      });
+    }
+    return intResult;
+  }
+
   @Override
   public int[] transformToIntValuesSV(ValueBlock valueBlock) {
     if (_resultMetadata.getDataType().getStoredType() != DataType.INT) {
       return super.transformToIntValuesSV(valueBlock);
     }
-    int[] selected = getSelectedArray(valueBlock);
+    int[] selected = getSelectedArray(valueBlock, false);
     int numDocs = valueBlock.getNumDocs();
     initIntValuesSV(numDocs);
-    int numElseThenStatements = _elseThenStatements.size();
-    for (int i = 0; i < numElseThenStatements; i++) {
+    int numThenStatements = _thenStatements.size();
+    BitSet unselectedDocs = new BitSet();
+    unselectedDocs.set(0, numDocs);
+    for (int i = 0; i < numThenStatements; i++) {
       if (_selections[i]) {
-        TransformFunction transformFunction = _elseThenStatements.get(i);
+        TransformFunction transformFunction = _thenStatements.get(i);
         int[] intValues = transformFunction.transformToIntValuesSV(valueBlock);
-        if (_numSelections == 1) {
-          System.arraycopy(intValues, 0, _intValuesSV, 0, numDocs);
-        } else {
-          for (int j = 0; j < numDocs; j++) {
-            if (selected[j] == i) {
-              _intValuesSV[j] = intValues[j];
-            }
+        for (int docId = unselectedDocs.nextSetBit(0); docId >= 0; docId = unselectedDocs.nextSetBit(docId + 1)) {
```

Comment:
Is there a way this can be converted to bit operations which are faster.

Reply:
I meant we compute all the "then statements" that need to be computed. 

(The previous comment implicitly meant that because it suggested to use a map. The map's size may be smaller than `numThenStatements`.)

Reply:
Can you elaborate on how we only iterate unselectedDocs once?

Reply:
```
BitSet unselectedDocs = new BitSet();
unselectedDocs.set(0, numDocs);
Map<Integer, int[]> thenStatementsIndexToValues = new HashMap<>();
for (int i = 0; i < numThenStatements; i++) {
  if (_computeThenStatements[i]) {
    thenStatementsIndexToValues.put(i, _thenStatements.get(i).transformToIntValuesSV(valueBlock));
  }
}
for (int docId = 0; docId < numDocs; docId++) {
  if (selected[docId] >= 0) {
    _intValuesSV[docId] = thenStatementsIndexToValues.get(selected[docId])[docId];
    unselectedDocs.clear(docId);
  }
}
```

Reply:
done

Reply:
You want to make the same change to other functions (e.g. `transformToIntValuesSVWithNull`) right?

Reply:
This comment is not addressed.